### PR TITLE
Exceptions handling

### PR DESCRIPTION
**Pull request by mmmpolishchuk**

### CODE REVIEW

File: OutOfSchool/OutOfSchool.Tests/Services/WorkshopServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+            optionsBuilder = new DbContextOptionsBuilder<OutOfSchoolDbContext>()
+                .UseInMemoryDatabase("TestOutOfSchoolDb");
+
+            context = new OutOfSchoolDbContext(optionsBuilder.Options);
+
+            context.Database.EnsureCreated();
+            SeedDb();
+        }
+
+        // [TearDown]
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.Tests/Services/WorkshopServiceTests.cs:46]: if this code is not needed, please delete it

File: OutOfSchool/OutOfSchool.Tests/Services/WorkshopServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        [TestCase(1)]
+        [Order(6)]
+        public async Task Delete_WhenIdIsValid_ShouldDeleteWorkshopEntityWithThisId(long id)
+        {
+            // Act
+            var countBeforeDeleting = service.GetAll().Result.Data.Count();
+
+            await service.Delete(id).ConfigureAwait(false);
+
+            var countAfterDeleting = await service.GetAll().ConfigureAwait(false);
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.Tests/Services/WorkshopServiceTests.cs:195]: looks like Count() call is missing here

File: OutOfSchool/OutOfSchool.WebApi/Extensions/ExceptionMiddlewareExtension.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+using Microsoft.AspNetCore.Diagnostics;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using OutOfSchool.WebApi.Models;
+
+namespace OutOfSchool.WebApi.Extensions
+{
+    public class ExceptionMiddlewareExtension
+    {
+        private readonly RequestDelegate _next;
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Extensions/ExceptionMiddlewareExtension.cs:15]: _next -> next
let's be consistent

File: OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/ErrorDto.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+﻿namespace OutOfSchool.WebApi.Models.ResultModel
+{
+    public class ErrorDto
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/ErrorDto.cs:3]: why do we need this class?

File: OutOfSchool/OutOfSchool.WebApi/Extensions/ResultExtension.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        /// <typeparam name="T">Type of transferred data</typeparam>
+        /// <param name="result">Result T variable with data or error to transfer to consumer from controller</param>
+        /// <returns>Status code depending Result T data. If ErrorData is not empty, return error status code,
+        /// and error message description in Json format.
+        /// If no error transferred, returns OkResult (200 or 204 status code)</returns>
+        public static ActionResult ToActionResult<T>(this Result<T> result)
+        {
+            if (Equals(result, null))
+            {
+                result = new Result<T>
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Extensions/ResultExtension.cs:20]: why do we need this assignment?

File: OutOfSchool/OutOfSchool.WebApi/Extensions/ResultExtension.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+                return new JsonResult(ToErrorDto(result.Error))
+                {
+                    StatusCode = 500,
+                };
+            }
+
+            if (result.Error != null)
+            {
+                switch (result.Error.Code)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Extensions/ResultExtension.cs:37]: do we need this switch?

File: OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        /// <returns>Return transferred data in success option, or error data.</returns>
+        /// <exception cref="ArgumentNullException">Exception thrown if transferred data is empty or null.</exception>
+        public static Result<T> GetSuccess(T transferredData)
+        {
+            
+            if (object.Equals(transferredData, default(T) ) && typeof(T) != typeof(bool)) // default(bool) is false
+            {
+                throw new ArgumentNullException();
+            }
+            else
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs:30]: else block is not needed, code will be a bit shorter

File: OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            }
+        }
+
+        /// <summary>
+        /// If followed code assume returning error, use Error method to return Error. 
+        /// It will convert data into Return T data type.
+        /// </summary>
+        public static Result<T> GetError(ErrorCode errorCode, string errorMessage)
+        {
+            var newResult = new Result<T>
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs:47]: could this variable be named 'result'?

File: OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    /// </summary>
+    /// <typeparam name="T">Generic type of data type transferred.</typeparam>
+    public class Result<T>
+    {
+        public T Data { get; set; }
+
+        public ErrorData Error { get; set; }
+
+        /// <summary>
+        /// If followed code assume returning data without errors, use Success method.
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs:16]: what does this method do?

File: OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
+﻿using System;
+
+namespace OutOfSchool.WebApi.Models.ResultModel
+{
+    /// <summary>
+    /// Used to transfer data or errors information through app layers. Methods has to return Result T data type
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Models/ResultModel/Result.cs:6]: Methods have to

File: OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs [hunk 11]
```diff
@@ -27,112 +28,77 @@ public ChildService(IEntityRepository<Child> repository)
...
-              
-                return child.ToModel();
+
+                return Result<ChildDTO>.GetSuccess(child.ToModel());
             }
-            catch (Exception ex)
+            catch
             {
-                throw new Exception($"{nameof(dto)} could not be updated: {ex.Message}");
+                return Result<ChildDTO>.GetError(ErrorCode.InternalServerError, "Internal server error.");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs:84]: is any logging supposed to be here?
we are loosing any exception details here

File: OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs [hunk 12]
```diff
@@ -27,112 +28,77 @@ public ChildService(IEntityRepository<Child> repository)
...
-            }
-
-            if (dto.MiddleName.Length == 0)
-            {
-                throw new ArgumentException("Empty middlename.", nameof(dto));
-            }
-
             try
             {
+                if (dto == null)
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs:73]: what about the rest of the checks?

File: OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs [hunk 13]
```diff
@@ -27,112 +28,77 @@ public ChildService(IEntityRepository<Child> repository)
...
+        public async Task<Result<long>> Delete(long id)
         {
             try
             {
-                await repository
-                    .Delete(await repository.GetById(id).ConfigureAwait(false))
-                    .ConfigureAwait(false);
+                var dtoToDelete = new ChildDTO() { Id = id };
+                
+                await repository.Delete(dtoToDelete.ToDomain()).ConfigureAwait(false);
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs:95]: could we have a Delete method that takes just an id as a parameter?

Comment (mmmpolishchuk) [reply OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs:95]: Do I need to change the implementation of Repository? If yes, I suppose it should be only in this way as shown on the screenshot.
image

Comment (mmmpolishchuk) [reply OutOfSchool/OutOfSchool.WebApi/Services/ChildService.cs:95]: then rest of the team must change it as well, and rewrite tests

**Pull request closed without merging**